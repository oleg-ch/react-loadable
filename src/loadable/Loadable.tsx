import { Capture } from './Capture'
import { LoadableMap } from './LoadableMap'
import { createLoadable, load, preloadAll, preloadReady } from './utils'
import { LoadableOptions } from '../types'

const Loadable = (opts: LoadableOptions) => {
    return createLoadable(load, opts)
}

Loadable.preloadAll = preloadAll
Loadable.preloadReady = preloadReady
Loadable.Map = LoadableMap
Loadable.Capture = Capture

export default Loadable
