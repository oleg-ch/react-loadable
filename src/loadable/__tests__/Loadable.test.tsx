import React from 'react'
import renderer from 'react-test-renderer'

import Loadable from '../Loadable'

const waitFor = (delay: any) => {
    return new Promise((resolve) => {
        setTimeout(resolve, delay)
    })
}

const createLoader = (delay: any, loader: any, error?: any) => {
    return () => {
        return waitFor(delay).then(() => {
            if (loader) {
                return loader()
            }

            throw new Error(error)
        })
    }
}

const MyLoadingComponent: React.ComponentType = (props: any) => {
    return <div>MyLoadingComponent {JSON.stringify(props)}</div>
}

const MyComponent: React.ComponentType = (props: any) => {
    return <div>MyComponent {JSON.stringify(props)}</div>
}

describe('Tests for Loadable', () => {
    afterEach(async () => {
        await Loadable.preloadAll()
    })

    it('Works fine under normal conditions', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(400, () => MyComponent),
            loading: MyLoadingComponent,
        })

        const component1 = renderer.create(<LoadableMyComponent prop="foo" />)

        expect(component1.toJSON()).toMatchSnapshot() // initial
        await waitFor(200)
        expect(component1.toJSON()).toMatchSnapshot() // loading
        await waitFor(200)
        expect(component1.toJSON()).toMatchSnapshot() // loaded

        const component2 = renderer.create(<LoadableMyComponent prop="bar" />)

        expect(component2.toJSON()).toMatchSnapshot() // reload
    })

    it('Can use delay and timeout as expected', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(300, () => MyComponent),
            loading: MyLoadingComponent,
            delay: 100,
            timeout: 200,
        })

        const component1 = renderer.create(<LoadableMyComponent prop="foo" />)

        expect(component1.toJSON()).toMatchSnapshot() // initial
        await waitFor(100)
        expect(component1.toJSON()).toMatchSnapshot() // loading
        await waitFor(100)
        expect(component1.toJSON()).toMatchSnapshot() // timed out
        await waitFor(100)
        expect(component1.toJSON()).toMatchSnapshot() // loaded
    })

    it('Renders with error passed to Loading component', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(400, null, 'test error message'),
            loading: MyLoadingComponent,
        })

        const component = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component.toJSON()).toMatchSnapshot() // initial
        await waitFor(200)
        expect(component.toJSON()).toMatchSnapshot() // loading
        await waitFor(200)
        expect(component.toJSON()).toMatchSnapshot() // errored
    })

    it('Can manually call preload', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(400, () => MyComponent),
            loading: MyLoadingComponent,
        })

        // @ts-ignore
        const promise = LoadableMyComponent.preload()
        await waitFor(200)

        const component1 = renderer.create(<LoadableMyComponent prop="foo" />)

        expect(component1.toJSON()).toMatchSnapshot() // still loading...
        await promise
        expect(component1.toJSON()).toMatchSnapshot() // success

        const component2 = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component2.toJSON()).toMatchSnapshot() // success
    })

    it('Can use provided render function', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(400, () => ({ MyComponent })),
            loading: MyLoadingComponent,
            render(loaded, props) {
                const { MyComponent: MyComp } = loaded

                return <MyComp {...props} />
            },
        })

        const component = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component.toJSON()).toMatchSnapshot() // initial
        await waitFor(200)
        expect(component.toJSON()).toMatchSnapshot() // loading
        await waitFor(200)
        expect(component.toJSON()).toMatchSnapshot() // success
    })

    it('Works fine with Loadable.Map under normal conditions', async () => {
        const LoadableMyComponent = Loadable.Map({
            loader: {
                a: createLoader(200, () => MyComponent),
                b: createLoader(400, () => MyComponent),
            },
            loading: MyLoadingComponent,
            render(loaded, props) {
                const { a: A, b: B } = loaded

                return (
                    <div>
                        <A {...props} />
                        <B {...props} />
                    </div>
                )
            },
        })

        const component = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component.toJSON()).toMatchSnapshot() // initial
        await waitFor(200)
        expect(component.toJSON()).toMatchSnapshot() // loading
        await waitFor(200)
        expect(component.toJSON()).toMatchSnapshot() // success
    })

    it('Loadable.Map renders Loading component with error', async () => {
        const LoadableMyComponent = Loadable.Map({
            loader: {
                a: createLoader(200, () => MyComponent),
                b: createLoader(400, null, 'test map error'),
            },
            loading: MyLoadingComponent,
            render(loaded, props) {
                const { a: A, b: B } = loaded

                return (
                    <div>
                        <A {...props} />
                        <B {...props} />
                    </div>
                )
            },
        })

        const component = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component.toJSON()).toMatchSnapshot() // initial
        await waitFor(200)
        expect(component.toJSON()).toMatchSnapshot() // loading
        await waitFor(200)
        expect(component.toJSON()).toMatchSnapshot() // success
    })
})

describe('Tests for preloadReady', () => {
    beforeEach(() => {
        Reflect.defineProperty(global, '__webpack_modules__', {
            value: { '1': true, '2': true },
            enumerable: true,
            configurable: true,
        })
    })

    afterEach(() => {
        Reflect.deleteProperty(global, '__webpack_modules__')
    })

    it('Works fine if webpack prop is undefined', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(200, () => MyComponent),
            loading: MyLoadingComponent,
        })

        await Loadable.preloadReady()

        const component = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component.toJSON()).toMatchSnapshot()
    })

    it('Preloads component if its chunk is available', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(200, () => MyComponent),
            loading: MyLoadingComponent,
            webpack: () => ['1'],
        })

        await Loadable.preloadReady()

        const component = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component.toJSON()).toMatchSnapshot()
    })

    it('Preloads component if all its chunks are available', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(200, () => MyComponent),
            loading: MyLoadingComponent,
            webpack: () => ['1', '2'],
        })

        await Loadable.preloadReady()

        const component = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component.toJSON()).toMatchSnapshot()
    })

    it('Renders Loading component when some chunks are missing', async () => {
        const LoadableMyComponent = Loadable({
            loader: createLoader(200, () => MyComponent),
            loading: MyLoadingComponent,
            webpack: () => ['1', '42'],
        })

        await Loadable.preloadReady()

        const component = renderer.create(<LoadableMyComponent prop="baz" />)

        expect(component.toJSON()).toMatchSnapshot()
    })
})

describe('Tests for Loadable.Capture', () => {
    it('Reports modules', async () => {
        // @ts-ignore
        const __webpack_modules__ = {
            './resolved/path/to/Main.js': () => {},
        }

        const Main = () => <p>MAIN COMPONENT</p>
        const Loading = () => <p>Loading...</p>
        // const chunk = { default: Main, __esModule: true }

        const MainAsync = Loadable({
            loading: Loading,
            loader: createLoader(200, () => Main),
            modules: ['./Main.js'],
            webpack: () => ['./resolved/path/to/Main.js'],
        })

        class App extends React.Component {
            render() {
                return (
                    <div>
                        <MainAsync />
                    </div>
                )
            }
        }

        const modules: any[] = []
        const report = (m: any) => {
            modules.push(m)
        }

        await Loadable.preloadAll()

        const Comp = () => (
            <Loadable.Capture report={report}>
                <App />
            </Loadable.Capture>
        )

        const component = renderer.create(<Comp />)

        expect(component.toJSON()).toMatchSnapshot()
        expect(modules.length).toEqual(1)
        expect(modules[0]).toEqual('./Main.js')
    })
})
