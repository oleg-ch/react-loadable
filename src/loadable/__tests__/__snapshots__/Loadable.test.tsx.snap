// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Loadable Can manually call preload 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Can manually call preload 2`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;

exports[`Tests for Loadable Can manually call preload 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`Tests for Loadable Can use delay and timeout as expected 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Can use delay and timeout as expected 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Can use delay and timeout as expected 3`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":true,"error":null}
</div>
`;

exports[`Tests for Loadable Can use delay and timeout as expected 4`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;

exports[`Tests for Loadable Can use provided render function 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Can use provided render function 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Can use provided render function 3`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`Tests for Loadable Loadable.Map renders Loading component with error 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Loadable.Map renders Loading component with error 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Loadable.Map renders Loading component with error 3`] = `
<div>
  MyLoadingComponent 
  {"isLoading":false,"pastDelay":true,"timedOut":false,"error":{}}
</div>
`;

exports[`Tests for Loadable Renders with error passed to Loading component 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Renders with error passed to Loading component 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Renders with error passed to Loading component 3`] = `
<div>
  MyLoadingComponent 
  {"isLoading":false,"pastDelay":true,"timedOut":false,"error":{}}
</div>
`;

exports[`Tests for Loadable Works fine under normal conditions 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Works fine under normal conditions 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Works fine under normal conditions 3`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;

exports[`Tests for Loadable Works fine under normal conditions 4`] = `
<div>
  MyComponent 
  {"prop":"bar"}
</div>
`;

exports[`Tests for Loadable Works fine with Loadable.Map under normal conditions 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Works fine with Loadable.Map under normal conditions 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for Loadable Works fine with Loadable.Map under normal conditions 3`] = `
<div>
  <div>
    MyComponent 
    {"prop":"baz"}
  </div>
  <div>
    MyComponent 
    {"prop":"baz"}
  </div>
</div>
`;

exports[`Tests for Loadable.Capture Reports modules 1`] = `
<div>
  <p>
    MAIN COMPONENT
  </p>
</div>
`;

exports[`Tests for preloadReady Preloads component if all its chunks are available 1`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`Tests for preloadReady Preloads component if its chunk is available 1`] = `
<div>
  MyComponent 
  {"prop":"baz"}
</div>
`;

exports[`Tests for preloadReady Renders Loading component when some chunks are missing 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`Tests for preloadReady Works fine if webpack prop is undefined 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;
